File: modules/app/agent-repl/daemon/internal/ssm/mergedequeue_test.go
package ssm

import (
	"testing"

	frontendv1 "agentrepl/proto/agentshim/frontend/v1"

	"claude-repld/internal/workspace/merge"
)

// Every case reads the offer off mustCurrent rather than off the map directly:
// the claim being made is that the ONE WorkspaceState construction funnel
// stamps it, so a test that read the retained value would pass with the funnel
// unwired and the card would never appear.

func waitingStanding() merge.Standing {
	return merge.Standing{Repo: "/repos/alpha/.git", RunID: "run-waiting", Position: 3, Depth: 5}
}

func headStanding() merge.Standing {
	return merge.Standing{Repo: "/repos/alpha/.git", RunID: "run-head", Position: 1, Depth: 2, Head: true}
}

// A workspace nobody interrupted carries NO offer. Absence is what tells a
// frontend there is no card to draw, so a zero-valued one standing in for it
// would put an unanswerable question on every workspace at once.
func TestNoDequeueOfferWithoutAnInterrupt(t *testing.T) {
	// Arrange.
	m, _, _ := openTest(t, fakeResolver{"s1": "ws1"})
	if err := applyTest(m, evSessionStarted("s1", 1)); err != nil {
		t.Fatalf("Apply: %v", err)
	}

	// Act.
	got := mustCurrent(t, m, "ws1")

	// Assert.
	if got.GetMergeDequeueOffer() != nil {
		t.Fatalf("merge_dequeue_offer = %v, want none for a workspace nobody asked about", got.GetMergeDequeueOffer())
	}
}

// A WAITING standing becomes the waiting arm, carrying how many merges are
// ahead of it — the fact the card's message turns on.
func TestARaisedOfferCarriesTheWaitingStanding(t *testing.T) {
	// Arrange.
	m, _, _ := openTest(t, fakeResolver{"s1": "ws1"})
	applyPhases(t, m, "ws1", merge.PhaseMergeQueued)

	// Act.
	if _, err := m.RaiseMergeDequeueOffer("ws1", waitingStanding()); err != nil {
		t.Fatalf("RaiseMergeDequeueOffer: %v", err)
	}

	// Assert.
	offer := mustDequeueOffer(t, m, "ws1")
	if offer.GetRunId() != "run-waiting" {
		t.Fatalf("run_id = %q, want run-waiting", offer.GetRunId())
	}
	waiting := offer.GetWaiting()
	if waiting == nil {
		t.Fatalf("offer standing = %T, want the waiting arm", offer.GetStanding())
	}
	if waiting.GetAhead() != 2 || waiting.GetPosition() != 3 || waiting.GetDepth() != 5 {
		t.Fatalf("waiting = ahead %d, %d of %d; want ahead 2, 3 of 5",
			waiting.GetAhead(), waiting.GetPosition(), waiting.GetDepth())
	}
	if offer.GetOfferId() == "" {
		t.Fatal("offer_id is empty; an offer nothing can name is a question the user cannot answer")
	}
	if offer.GetRaisedAtMs() == 0 {
		t.Fatal("raised_at_ms is zero, want the instant the question went up")
	}
}

// A HEAD standing becomes the running arm, carrying the run's own MergeStatus
// so the card names the stage from the one vocabulary that defines it.
func TestARaisedOfferCarriesTheRunningStatus(t *testing.T) {
	// Arrange — a run mid-cherry-pick.
	m, _, _ := openTest(t, fakeResolver{"s1": "ws1"})
	if err := m.ApplyMergeStatus("ws1", string(merge.PhaseMerging), "cherry-picking 3/7",
		pipelineStatus("run-head", 7, 3, "abc123def456", "fix the parser")); err != nil {
		t.Fatalf("ApplyMergeStatus: %v", err)
	}

	// Act.
	if _, err := m.RaiseMergeDequeueOffer("ws1", headStanding()); err != nil {
		t.Fatalf("RaiseMergeDequeueOffer: %v", err)
	}

	// Assert.
	offer := mustDequeueOffer(t, m, "ws1")
	running := offer.GetRunning()
	if running == nil {
		t.Fatalf("offer standing = %T, want the running arm", offer.GetStanding())
	}
	picking := running.GetStatus().GetCherryPicking()
	if picking == nil {
		t.Fatalf("running status phase = %T, want the cherry-picking phase the run published", running.GetStatus().GetPhase())
	}
	if picking.GetCommitsLanded() != 3 || picking.GetCommitsTotal() != 7 {
		t.Fatalf("running status = %d/%d commits, want 3/7", picking.GetCommitsLanded(), picking.GetCommitsTotal())
	}
}

// A head observed before its run has published anything carries the running arm
// with NO status. It is not an error and not a placeholder: the merge really is
// running and really has said nothing yet, and the arm alone is enough for the
// card to say so.
func TestARaisedOfferOnASilentHeadCarriesTheArmWithoutAStatus(t *testing.T) {
	// Arrange — the merge axis says merging, but no run has published.
	m, _, _ := openTest(t, fakeResolver{"s1": "ws1"})
	applyPhases(t, m, "ws1", merge.PhaseMerging)

	// Act.
	if _, err := m.RaiseMergeDequeueOffer("ws1", headStanding()); err != nil {
		t.Fatalf("RaiseMergeDequeueOffer: %v", err)
	}

	// Assert.
	offer := mustDequeueOffer(t, m, "ws1")
	if offer.GetRunning() == nil {
		t.Fatalf("offer standing = %T, want the running arm", offer.GetStanding())
	}
	if offer.GetRunning().GetStatus() != nil {
		t.Fatalf("running status = %v, want none from a run that has published nothing", offer.GetRunning().GetStatus())
	}
}

// A SECOND raise keeps the offer id and refreshes the standing. Interrupting
// twice is ordinary, and a rival question would leave two cards whose answers
// race; the refresh is also what keeps a card honest as the queue advances.
func TestASecondRaiseRefreshesTheStandingAndKeepsTheOfferID(t *testing.T) {
	// Arrange.
	m, _, _ := openTest(t, fakeResolver{"s1": "ws1"})
	applyPhases(t, m, "ws1", merge.PhaseMergeQueued)
	if _, err := m.RaiseMergeDequeueOffer("ws1", waitingStanding()); err != nil {
		t.Fatalf("RaiseMergeDequeueOffer(first): %v", err)
	}
	first := mustDequeueOffer(t, m, "ws1").GetOfferId()

	// Act — the merge has advanced to second in line.
	advanced := waitingStanding()
	advanced.Position, advanced.Depth = 2, 4
	if _, err := m.RaiseMergeDequeueOffer("ws1", advanced); err != nil {
		t.Fatalf("RaiseMergeDequeueOffer(second): %v", err)
	}

	// Assert.
	offer := mustDequeueOffer(t, m, "ws1")
	if offer.GetOfferId() != first {
		t.Fatalf("offer_id = %q after a second raise, want the original %q", offer.GetOfferId(), first)
	}
	if got := offer.GetWaiting().GetAhead(); got != 1 {
		t.Fatalf("ahead = %d after the refresh, want 1", got)
	}
}

// Clearing takes the card down, which is the only way it comes down.
func TestClearingTheOfferTakesTheCardDown(t *testing.T) {
	// Arrange.
	m, _, _ := openTest(t, fakeResolver{"s1": "ws1"})
	applyPhases(t, m, "ws1", merge.PhaseMergeQueued)
	if _, err := m.RaiseMergeDequeueOffer("ws1", waitingStanding()); err != nil {
		t.Fatalf("RaiseMergeDequeueOffer: %v", err)
	}

	// Act.
	had, err := m.ClearMergeDequeueOffer("ws1", "answered:keep")

	// Assert.
	if err != nil || !had {
		t.Fatalf("ClearMergeDequeueOffer() = (%v, %v), want (true, nil)", had, err)
	}
	if got := mustCurrent(t, m, "ws1").GetMergeDequeueOffer(); got != nil {
		t.Fatalf("merge_dequeue_offer = %v after the clear, want none", got)
	}
}

// Clearing a workspace with no question is a benign no-op, not an error: an
// answer racing a merge that ended on its own is an ordinary sequence.
func TestClearingAnAbsentOfferIsANoOp(t *testing.T) {
	// Arrange.
	m, _, _ := openTest(t, fakeResolver{"s1": "ws1"})

	// Act.
	had, err := m.ClearMergeDequeueOffer("ws1", "answered:keep")

	// Assert.
	if err != nil || had {
		t.Fatalf("ClearMergeDequeueOffer() = (%v, %v), want (false, nil)", had, err)
	}
}

// THE MERGE'S OWN END ANSWERS THE QUESTION, one row per terminal. A card
// offering to dequeue a merge that is already gone is a question about nothing.
func TestAMergeReachingItsOwnEndTakesTheCardDown(t *testing.T) {
	tests := []struct {
		name  string
		token string
	}{
		{"merged", sigMerged},
		{"failed", sigMergeFailed},
		{"axis cleared", sigMergeNone},
	}
	for _, tc := range tests {
		t.Run(tc.name, func(t *testing.T) {
			// Arrange — a session row too, so the workspace still resolves
			// once the merge axis is cleared and the frame can be read back.
			m, _, _ := openTest(t, fakeResolver{"s1": "ws1"})
			if err := applyTest(m, evSessionStarted("s1", 1)); err != nil {
				t.Fatalf("Apply: %v", err)
			}
			applyPhases(t, m, "ws1", merge.PhaseMergeQueued)
			if _, err := m.RaiseMergeDequeueOffer("ws1", waitingStanding()); err != nil {
				t.Fatalf("RaiseMergeDequeueOffer: %v", err)
			}

			// Act.
			if err := m.ApplyMergeTransition("ws1", tc.token, "test arrangement"); err != nil {
				t.Fatalf("ApplyMergeTransition(%s): %v", tc.token, err)
			}

			// Assert.
			if got := mustCurrent(t, m, "ws1").GetMergeDequeueOffer(); got != nil {
				t.Fatalf("merge_dequeue_offer = %v after %s, want none", got, tc.token)
			}
			if _, ok := m.MergeDequeueOfferID("ws1"); ok {
				t.Fatalf("an offer is still outstanding after %s, want none", tc.token)
			}
		})
	}
}

// A merge transition that is NOT an ending leaves the question standing: the
// merge is still on the queue, so the offer is still about something.
func TestANonTerminalMergeTransitionLeavesTheCardUp(t *testing.T) {
	// Arrange.
	m, _, _ := openTest(t, fakeResolver{"s1": "ws1"})
	applyPhases(t, m, "ws1", merge.PhaseMergeQueued)
	if _, err := m.RaiseMergeDequeueOffer("ws1", waitingStanding()); err != nil {
		t.Fatalf("RaiseMergeDequeueOffer: %v", err)
	}

	// Act — the merge reached the head and started picking.
	applyPhases(t, m, "ws1", merge.PhaseMerging)

	// Assert.
	if got := mustCurrent(t, m, "ws1").GetMergeDequeueOffer(); got == nil {
		t.Fatal("merge_dequeue_offer is absent after a non-terminal transition, want the question still standing")
	}
}

// The retained offer rides EVERY later frame, not just the one its raise
// pushed. A card that vanished on the next unrelated push would flicker off a
// question nobody answered.
func TestTheRetainedOfferRidesEveryLaterFrame(t *testing.T) {
	// Arrange.
	m, _, _ := openTest(t, fakeResolver{"s1": "ws1"})
	applyPhases(t, m, "ws1", merge.PhaseMergeQueued)
	if _, err := m.RaiseMergeDequeueOffer("ws1", waitingStanding()); err != nil {
		t.Fatalf("RaiseMergeDequeueOffer: %v", err)
	}

	// Act — an unrelated push for the same workspace.
	if err := applyTest(m, evSessionStarted("s1", 2)); err != nil {
		t.Fatalf("Apply: %v", err)
	}

	// Assert.
	if got := mustDequeueOffer(t, m, "ws1").GetRunId(); got != "run-waiting" {
		t.Fatalf("run_id = %q on a later frame, want the retained run-waiting", got)
	}
}

// MergeDequeueOfferID is what an answer is checked against, so it must report
// exactly what was published and nothing when nothing is up.
func TestMergeDequeueOfferIDReportsTheOutstandingQuestion(t *testing.T) {
	// Arrange.
	m, _, _ := openTest(t, fakeResolver{"s1": "ws1"})
	applyPhases(t, m, "ws1", merge.PhaseMergeQueued)
	if _, ok := m.MergeDequeueOfferID("ws1"); ok {
		t.Fatal("an offer id is reported before anything was raised")
	}
	if _, err := m.RaiseMergeDequeueOffer("ws1", waitingStanding()); err != nil {
		t.Fatalf("RaiseMergeDequeueOffer: %v", err)
	}

	// Act.
	got, ok := m.MergeDequeueOfferID("ws1")

	// Assert.
	if !ok {
		t.Fatal("MergeDequeueOfferID reported nothing outstanding after a raise")
	}
	if want := mustDequeueOffer(t, m, "ws1").GetOfferId(); got != want {
		t.Fatalf("MergeDequeueOfferID = %q, want the published %q", got, want)
	}
}

// An unnamed workspace is a construction bug at the call site, refused rather
// than filed under the empty string.
func TestRaisingAnOfferForAnEmptyWorkspaceIsRefused(t *testing.T) {
	// Arrange.
	m, _, _ := openTest(t, fakeResolver{"s1": "ws1"})

	// Act.
	offer, err := m.RaiseMergeDequeueOffer("", waitingStanding())

	// Assert.
	if err == nil || offer != nil {
		t.Fatalf("RaiseMergeDequeueOffer(\"\") = (%v, %v), want (nil, error)", offer, err)
	}
}

// mustDequeueOffer returns the workspace's stamped dequeue offer, failing when
// the frame carries none.
func mustDequeueOffer(t *testing.T, m *Manager, ws string) *frontendv1.MergeDequeueOffer {
	t.Helper()
	offer := mustCurrent(t, m, ws).GetMergeDequeueOffer()
	if offer == nil {
		t.Fatalf("WorkspaceState for %s carries no merge_dequeue_offer", ws)
	}
	return offer
}

File: modules/app/agent-repl/webapp/test/merge-dequeue.test.ts
/**
 * merge-dequeue — the card an interrupt raises over a queued merge.
 *
 * Its whole job is to say WHERE the merge stands and to make the destructive
 * answer deliberate, so these assert the standing arm reaching the prose, the
 * card being exactly as present as the pushed offer, and the answer reaching
 * the socket as itself. One edge per test.
 */
import { describe, expect, it } from "vitest";

import type { MergeDequeueOffer, MergeStatus } from "../src/frontend-proto.js";
import {
  DEQUEUE_ATTR,
  DEQUEUE_HEADING,
  dequeueAnswerFromAttr,
  dequeueOfferText,
  dequeueRefusedNotice,
  dequeueStageText,
  mergeDequeueCardHtml,
} from "../src/merge-dequeue.js";

function waitingOffer(ahead: number, position: number, depth: number): MergeDequeueOffer {
  return {
    offerId: "offer-1",
    runId: "run-7",
    raisedAtMs: 1000,
    standing: { case: "waiting", value: { ahead, position, depth } },
  };
}

function runningOffer(status?: MergeStatus): MergeDequeueOffer {
  return {
    offerId: "offer-1",
    runId: "run-7",
    raisedAtMs: 1000,
    standing: { case: "running", value: status === undefined ? {} : { status } },
  };
}

function statusWith(phase: MergeStatus["phase"]): MergeStatus {
  return { runId: "run-7", phaseStartedAtMs: 1, updatedAtMs: 2, phase };
}

describe("no outstanding question", () => {
  // The empty string IS the absent card: the caller writes this into its slot
  // every frame, so a cleared offer takes the card down by the same path that
  // put it up.
  it("draws nothing at all", () => {
    expect(mergeDequeueCardHtml(null)).toBe("");
  });
});

describe("a waiting merge", () => {
  // The count is the fact that made the user press the key: "how much longer".
  it("says how many merges are ahead of it", () => {
    expect(dequeueOfferText(waitingOffer(2, 3, 5))).toContain("with 2 merges ahead of it");
  });

  it("says where it sits on the queue", () => {
    expect(dequeueOfferText(waitingOffer(2, 3, 5))).toContain("It is 3 of 5 on the queue");
  });

  // Singular, because "with 1 merges ahead of it" is the kind of seam that
  // makes a reader distrust the number beside it.
  it("says one merge ahead in the singular", () => {
    expect(dequeueOfferText(waitingOffer(1, 2, 4))).toContain("with 1 merge ahead of it");
  });

  // The reassurance is the point of the waiting arm: nothing has run, so the
  // destructive button is not destructive here.
  it("says taking it off costs nothing", () => {
    expect(dequeueOfferText(waitingOffer(2, 3, 5))).toContain("costs nothing");
  });
});

describe("a running merge", () => {
  // The stage is what the running arm exists to say, and it comes from the
  // run's own MergeStatus rather than from a stage word the card invented.
  it("names the cherry-pick it is on", () => {
    const offer = runningOffer(
      statusWith({
        case: "cherryPicking",
        value: {
          commitsTotal: 7,
          commitsLanded: 3,
          currentSha: "abc123",
          currentSubject: "fix the parser",
        },
      }),
    );
    expect(dequeueOfferText(offer)).toContain("cherry-picking commit 4 of 7");
  });

  it("names the test gate it is on", () => {
    const offer = runningOffer(
      statusWith({
        case: "testing",
        value: {
          commitsTotal: 7,
          commitsLanded: 5,
          currentSha: "abc123",
          currentSubject: "fix the parser",
        },
      }),
    );
    expect(dequeueOfferText(offer)).toContain("testing 5 of 7 landed commits");
  });

  it("names a parked conflict", () => {
    const offer = runningOffer(
      statusWith({
        case: "conflict",
        value: {
          conflictedSha: "abc123",
          conflictedSubject: "fix the parser",
          commitsTotal: 7,
          commitsLanded: 3,
        },
      }),
    );
    expect(dequeueOfferText(offer)).toContain("parked on a conflict");
  });

  // A head that has published nothing yet is a real state, not a gap: the arm
  // alone is enough to say the merge is running.
  it("still says the merge is running with no status published", () => {
    expect(dequeueOfferText(runningOffer())).toContain("This merge is running");
  });

  // THE WARNING IS THE POINT OF THE RED. The one case where an abort is not
  // free is unrecoverable, so the card never implies it is.
  it("warns that anything already landed stays landed", () => {
    expect(dequeueOfferText(runningOffer())).toContain(
      "anything it has already landed on the target stays there",
    );
  });

  it("says nothing about a stage for a status that has not been published", () => {
    expect(dequeueStageText(undefined)).toBe("");
  });
});

describe("the card", () => {
  it("carries the heading naming the verb", () => {
    expect(mergeDequeueCardHtml(waitingOffer(2, 3, 5))).toContain(DEQUEUE_HEADING);
  });

  it("offers both answers as buttons", () => {
    const html = mergeDequeueCardHtml(waitingOffer(2, 3, 5));
    expect(html).toContain(`${DEQUEUE_ATTR}="dequeue"`);
    expect(html).toContain(`${DEQUEUE_ATTR}="keep"`);
  });

  // The class is what the stylesheet's red frame hangs off; a card that lost it
  // would ask a destructive question dressed as an ordinary one.
  it("marks a running offer so the frame can differ from a waiting one", () => {
    expect(mergeDequeueCardHtml(runningOffer())).toContain('class="merge-dequeue running"');
    expect(mergeDequeueCardHtml(waitingOffer(1, 2, 2))).toContain(
      'class="merge-dequeue waiting"',
    );
  });
});

describe("reading the clicked answer", () => {
  it("reads dequeue", () => {
    expect(dequeueAnswerFromAttr("dequeue")).toBe("dequeue");
  });

  it("reads keep", () => {
    expect(dequeueAnswerFromAttr("keep")).toBe("keep");
  });

  // "Not the keep button" is not the same statement as "dequeue" when one of
  // the two answers destroys work, so an unrecognized value throws rather than
  // resolving to whichever the reader guessed.
  it("throws on anything else rather than guessing", () => {
    expect(() => dequeueAnswerFromAttr("maybe")).toThrow(/unrecognized answer/);
  });

  it("throws on a missing attribute", () => {
    expect(() => dequeueAnswerFromAttr(null)).toThrow(/unrecognized answer/);
  });
});

describe("a refused answer", () => {
  // The card comes down only when the daemon clears the offer, so a refusal
  // rendered nowhere would read as the button doing nothing.
  it("names the verb that was refused and the daemon's own words", () => {
    const notice = dequeueRefusedNotice("dequeue", new Error("no dequeue offer is outstanding"));
    expect(notice).toBe(
      "could not dequeue this merge: no dequeue offer is outstanding",
    );
  });

  it("names the keep verb for a refused keep", () => {
    expect(dequeueRefusedNotice("keep", new Error("offer is stale"))).toContain(
      "could not keep this merge queued",
    );
  });
});
